The "cannot create non-exhaustive struct using struct expression" warning should silence the "cannot construct `Foo` with struct literal syntax due to private fields" warning

### Code

```Rust
// src/lib.rs
#[non_exhaustive]
pub struct Foo {
    pub my_field: u32,
    private_field: i32,
}
```
```Rust
// src/main.rs
use exhaustive_test::Foo;

fn main() {
    let foo = Foo {
        my_field: 10,
    };
}
```

### Current output

```text
error[E0639]: cannot create non-exhaustive struct using struct expression
 --> src/main.rs:4:15
  |
4 |       let foo = Foo {
  |  _______________^
5 | |         my_field: 10,
6 | |     };
  | |_____^

error: cannot construct `Foo` with struct literal syntax due to private fields
 --> src/main.rs:4:15
  |
4 |     let foo = Foo {
  |               ^^^
  |
  = note: ...and other private field `private_field` that was not provided
```

### Desired output

```text
error[E0639]: cannot create non-exhaustive struct using struct expression
 --> src/main.rs:4:15
  |
4 |       let foo = Foo {
  |  _______________^
5 | |         my_field: 10,
6 | |     };
  | |_____^
```

### Rationale and extra context

There is no reason to provide two errors for the same issue, and the `#[non_exhaustive]` issue is the one that should be highlighted since it's explicitly chosen by the struct author.

### Other cases

If there is no private field, then the correct output is created.

### Rust Version

```text
$ rustc +nightly --version --verbose
rustc 1.92.0-nightly (b6f0945e4 2025-10-08)
binary: rustc
commit-hash: b6f0945e4681bc4d2faa7c22c5f61dc36abf7dd2
commit-date: 2025-10-08
host: x86_64-unknown-linux-gnu
release: 1.92.0-nightly
LLVM version: 21.1.2
```

### Anything else?

_No response_

--- a/compiler/rustc_hir_typeck/src/expr.rs
+++ b/compiler/rustc_hir_typeck/src/expr.rs
@@ -2331,7 +2331,11 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 }
             };
             self.typeck_results.borrow_mut().fru_field_types_mut().insert(expr.hir_id, fru_tys);
-        } else if adt_kind != AdtKind::Union && !remaining_fields.is_empty() {
+        } else if adt_kind != AdtKind::Union
+            && !remaining_fields.is_empty()
+            //~ non_exhaustive already reported, which will only happen for extern modules
+            && !variant.field_list_has_applicable_non_exhaustive()
+        {
             debug!(?remaining_fields);
             let private_fields: Vec<&ty::FieldDef> = variant
                 .fields

--- a/tests/ui/privacy/auxiliary/non_exhaustive_with_private.rs
+++ b/tests/ui/privacy/auxiliary/non_exhaustive_with_private.rs
@@ -1,0 +1,13 @@
+// Auxiliary crate for testing non-exhaustive struct with private fields
+
+#[non_exhaustive]
+pub struct Foo {
+    pub my_field: u32,
+    private_field: i32,
+}
+
+#[non_exhaustive]
+pub struct Bar {
+    pub my_field: u32,
+    pub missing_field: i32,
+}

--- a/tests/ui/privacy/non-exhaustive-with-private-fields-147513.rs
+++ b/tests/ui/privacy/non-exhaustive-with-private-fields-147513.rs
@@ -1,0 +1,17 @@
+//@ aux-build:non_exhaustive_with_private.rs
+
+extern crate non_exhaustive_with_private;
+
+use non_exhaustive_with_private::{Bar, Foo};
+
+fn main() {
+    let foo = Foo {
+        //~^ ERROR cannot create non-exhaustive struct using struct expression
+        my_field: 10,
+    };
+
+    let bar = Bar {
+        //~^ ERROR cannot create non-exhaustive struct using struct expression
+        my_field: 10,
+    };
+}

--- a/tests/ui/privacy/non-exhaustive-with-private-fields-147513.stderr
+++ b/tests/ui/privacy/non-exhaustive-with-private-fields-147513.stderr
@@ -1,0 +1,23 @@
+error[E0639]: cannot create non-exhaustive struct using struct expression
+  --> $DIR/non-exhaustive-with-private-fields-147513.rs:8:15
+   |
+LL |       let foo = Foo {
+   |  _______________^
+LL | |
+LL | |         my_field: 10,
+LL | |     };
+   | |_____^
+
+error[E0639]: cannot create non-exhaustive struct using struct expression
+  --> $DIR/non-exhaustive-with-private-fields-147513.rs:13:15
+   |
+LL |       let bar = Bar {
+   |  _______________^
+LL | |
+LL | |         my_field: 10,
+LL | |     };
+   | |_____^
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0639`.

